<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.550">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>R 4 biostats - Course Description &amp; Objectives</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<link href="./images/icons8_hist_96.png" rel="icon" type="image/png">
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-51QEQ3S83R"></script>

<script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-51QEQ3S83R', { 'anonymize_ip': true});
</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">R 4 biostats</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> <i class="bi bi-house-fill" role="img">
</i> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./install.html"> <i class="bi bi-download" role="img">
</i> 
<span class="menu-text">Pre-Workshop</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./syllabus.html" aria-current="page"> <i class="bi bi-book-fill" role="img">
</i> 
<span class="menu-text">Syllabus</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-lectures" role="button" data-bs-toggle="dropdown" aria-expanded="false">
      <i class="bi bi-easel-fill" role="img">
</i> 
 <span class="menu-text">Lectures</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-lectures">    
        <li>
    <a class="dropdown-item" href="./lectures/lec01_data_with_R.html">
 <span class="dropdown-text">Intro to R and data analysis</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./lectures/lec02_inference.html">
 <span class="dropdown-text">Statistical inference &amp; hypothesis testing</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./lectures/lec03_corr_regress.html">
 <span class="dropdown-text">Modeling correlation and regression</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./lectures/lec04_intro_ML.html">
 <span class="dropdown-text">Intro to machine learning</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-fa-brands-r-project--labs" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text"><i class="fa-brands fa-r-project" aria-label="r-project"></i> Labs</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-fa-brands-r-project--labs">    
        <li>
    <a class="dropdown-item" href="./practice/lab01_data_with_R.html">
 <span class="dropdown-text">Lab: Intro to R and data analysis</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./practice/lab02_inference.html">
 <span class="dropdown-text">Lab: Statistical inference &amp; hypothesis testing</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./practice/lab03_corr_regress.html">
 <span class="dropdown-text">Lab: Modeling correlation and regression</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./practice/lab04_intro_ML.html">
 <span class="dropdown-text">Lab: Intro to machine learning</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="./me.html"> <i class="bi bi-mortarboard-fill" role="img">
</i> 
<span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./license_etc.html"> <i class="bi bi-bookmark-heart-fill" role="img">
</i> 
<span class="menu-text">Acknowledgements</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul class="collapse">
  <li><a href="#course-description" id="toc-course-description" class="nav-link active" data-scroll-target="#course-description">Course Description</a></li>
  <li><a href="#target-audience" id="toc-target-audience" class="nav-link" data-scroll-target="#target-audience">Target Audience</a></li>
  <li><a href="#prerequisites" id="toc-prerequisites" class="nav-link" data-scroll-target="#prerequisites">Prerequisites</a></li>
  <li><a href="#course-outline" id="toc-course-outline" class="nav-link" data-scroll-target="#course-outline">Course Outline</a></li>
  </ul>
</nav>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Course Description &amp; Objectives</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="course-description" class="level1">
<h1>Course Description</h1>
<!-- https://bioinformatics.ca/workshops-all/2024-help-what-statistical-model-should-i-use-guelph-on/ -->
<!-- This is a specialized course designed for bioinformatics professionals and researchers dealing with biological datasets who are unsure about which statistical analysis they should implement. This course focuses on empowering participants to conduct end-to-end data analyses. Learn the art of selecting appropriate statistical methods aligned with research questions, and develop a decision tree guiding data cleaning, exploration, analysis, and interpretation. Engage in hands-on coding sessions in the R software with real-world datasets, learning the skills needed to confidently navigate and communicate insights throughout the analytical process. Particular emphasis will be placed on developing an understanding of the statistical methods used, when to apply them, and how to interpret them. By the course’s conclusion, participants will be equipped with a concise and powerful toolkit for unlocking the full potential of complex biological datasets. -->
<p>This is an intensive course originally designed for PhD students and researchers dealing with biological datasets to help them consolidate their practical understanding of frequently implemented data analysis and statistical modeling for academic research purposes. This course focuses on empowering participants to conduct <strong>end-to-end data analyses</strong>, guiding them through reproducible data storage, cleaning, exploration, analysis, and interpretation cycle.</p>
<p>Beyond the <strong>conceptual foundations</strong>, participants engage in <strong>hands-on coding sessions</strong> in the R software with real-world datasets and solved questions, validating in practice the skills needed to confidently acquire and communicate insights throughout the analytical process.</p>
<p>Particular emphasis is placed on developing an understanding of the statistical methods used, when to apply them, and how to interpret them, in close connection to “real life” situations for a research scientists. <!-- By the course’s conclusion, participants will be equipped with a concise and powerful toolkit for unlocking the full potential of complex biological datasets. --></p>
<!-- # Course Objectives -->
<!-- Participants will gain practical experience and skills to be able to: -->
<!--   + Understand the challenges and opportunities presented by complex biological data. -->
<!--   + Apply fundamental techniques for exploring and summarizing biological datasets. -->
<!--   + Address missing values through sophisticated imputation methodologies.  -->
<!--   + Analyze relationships within high-dimensional data, both between variables and samples, employing regression methods, regularization techniques, and clustering methods. -->
<!--   + Integrate various statistical methods into a cohesive workflow to tackle a variety of common problems in bioinformatics from data exploration to interpretation. -->
</section>
<section id="target-audience" class="level1">
<h1>Target Audience</h1>
<p>The target audience for this course is graduate students, post-doctoral fellows, and researchers in industry or academia already familiar with basic statistics (ideally in R) and looking to learn more about conducting an analysis from start to finish with more intermediate or advanced statistical techniques.</p>
</section>
<section id="prerequisites" class="level1">
<h1>Prerequisites</h1>
<!-- Participants are expected to have completed the ‘Introduction to R’ and ‘Analysis Using R’ courses available through Bioinformatics.ca or possess equivalent proficiency in R. Previous years’ R workshop materials are available open-access here. In particular, we expect participants to feel comfortable with the following skills in R as our course will build upon them: -->
<!-- + Reading data into R -->
<!-- + Data types and classes  -->
<!-- + Manipulating data in R with Tidyverse (e.g., select, mutate, filter, etc.) -->
<!-- + Making plots in ggplot2 -->
<!-- + Writing custom functions -->
<!-- + Estimating a correlation matrix -->
<!-- + Hierarchical and K-means clustering algorithms -->
<!-- + Fitting a linear regression model with the lm function -->
<!-- + Fitting a logistic regression model with the glm function -->
<!-- You will also require your own laptop computer. Minimum requirements: 1024×768 screen resolution, 1.5GHz CPU, 2GB RAM, 10GB free disk space, recent versions of Windows, Mac OS X or Linux (Most computers purchased in the past 3-4 years likely meet these requirements). -->
<!-- This workshop requires participants to complete pre-workshop tasks and readings. -->
<p>While there are no specific prerequisites, because of the intensive nature of this course, participants will make the most of it if they have prior exposure to quantitative work in the biology/life science field. Similarly, some prior exposure to R programming will make the lab sessions more engaging.</p>
<p>Detailed <a href="https://r4biostats.com/install.html">instructions</a> are provided to complete the required installation of R and RStudio ahead of the workshop, so that participants can follow along the lab sessions.</p>
</section>
<section id="course-outline" class="level1">
<h1>Course Outline</h1>
<section id="module-1-introduction-to-r-and-data-analysis" class="level3">
<h3 class="anchored" data-anchor-id="module-1-introduction-to-r-and-data-analysis">Module 1: Introduction to R and data analysis</h3>
<ul>
<li><strong>Introduction to reproducible end-to-end analysis using R</strong>
<ul>
<li>Why use R?</li>
<li>Principles of reproducible analysis with R + RStudio</li>
<li>R objects, functions, packages</li>
</ul></li>
<li><strong>Discussion of different variable types (qualitative, quantitative) and levels of measurement (nominal, ordinal, interval, ratio)</strong>
<ul>
<li>Principles of “tidy data”</li>
<li>Introduction to data cleaning and manipulation methods</li>
</ul></li>
<li><strong>Descriptive statistics</strong>
<ul>
<li>Univariate analysis</li>
<li>Measures of central tendency, measures of variability (or spread), and frequency distributions</li>
</ul></li>
<li><strong>Visual data exploration</strong>
<ul>
<li>Introduction to <code>ggplot2</code> package for making graphs in R</li>
</ul></li>
</ul>
<!-- + Discussion of the importance of “tidy data” when performing statistical analyses -->
<!-- + Introduction to advanced data cleaning methods -->
<!-- + Discussion of methods from the perspective of finding patterns in order to plan out your analyses and pre-processing steps -->
</section>
<section id="module-2-statistical-inference-and-classical-hypothesis-testing" class="level3">
<h3 class="anchored" data-anchor-id="module-2-statistical-inference-and-classical-hypothesis-testing">Module 2: Statistical inference and classical hypothesis testing</h3>
<ul>
<li><strong>Purpose and foundations of inferential statistics</strong>
<ul>
<li>Probability and random variables</li>
<li>Meaningful probability distributions</li>
<li>Sampling distributions and Central Limit Theorem</li>
</ul></li>
<li><strong>Getting to know the “language” of hypothesis testing</strong>
<ul>
<li>The <em>null</em> and alternative hypothesis</li>
<li>The probability of error? (<em>α</em> or “significance level”)</li>
<li>The <em>p-value</em> probability and tests interpretation</li>
<li>Confidence Intervals</li>
<li>Types of errors (Type 1 and Type 2)</li>
<li>Effective vs statistical significance</li>
</ul></li>
<li><strong>Hypothesis tests examples</strong>
<ul>
<li>Comparing sample mean to a hypothesized population mean (Z test &amp; t test)</li>
<li>Comparing two independent sample means (t test)</li>
<li>Comparing sample means from 3 or more groups (ANOVA)</li>
</ul></li>
<li><strong>A closer look at testing assumptions (with examples)</strong>
<ul>
<li>Testing two groups that are <em>not</em> independent</li>
<li>Testing if the data are <em>not</em> normally distributed: non-parametric tests</li>
<li>Testing samples <em>without</em> homogeneous variance of observations</li>
</ul></li>
</ul>
</section>
<section id="module-3-modeling-correlation-and-regression" class="level3">
<h3 class="anchored" data-anchor-id="module-3-modeling-correlation-and-regression">Module 3: Modeling correlation and regression</h3>
<ul>
<li>Testing and summarizing relationship between 2 variables (<strong>correlation</strong>)
<ul>
<li>Pearson <span class="math inline">\(r\)</span> analysis (parametric)
<ul>
<li>(numerical variables)</li>
</ul></li>
<li>Spearman’s test (not parametric)</li>
</ul></li>
<li>Measures of <strong>association</strong>
<ul>
<li>Chi-Square Test of Independence
<ul>
<li>(categorical variables)</li>
</ul></li>
<li>Fisher’s Exact Test</li>
</ul></li>
<li>From correlation/association to <strong>prediction/causation</strong>
<ul>
<li>The purpose of observational and experimental studies</li>
</ul></li>
<li>Introduction of <strong>regression based statistical methods</strong>
<ul>
<li>Simple linear regression models</li>
<li>Multiple Linear Regression models</li>
</ul></li>
<li>Shifting the emphasis on <strong>empirical prediction</strong>
<ul>
<li>Introduction to Machine Learning (ML)</li>
<li>Distinction between Supervised and Unsupervised algorithms</li>
</ul></li>
</ul>
</section>
<section id="module-4-introduction-to-machine-learning" class="level3">
<h3 class="anchored" data-anchor-id="module-4-introduction-to-machine-learning">Module 4: Introduction to machine learning</h3>
<ul>
<li>Examples of Machine Learning algorithms
<ul>
<li><strong>PCA</strong> – “unsupervised” ML algorithm for dimensionality reduction</li>
<li><strong>PLS-Discriminant Analysis</strong> – “supervised” alternative to PCA performing simultaneous dimensionality reduction and classification <!-- + Shifting the emphasis on **empirical prediction**  --> <!--   + Distinction between supervised & **unsupervised** algorithms --> <!--     + Unsupervised ML Example --> <!--       + PCA  --></li>
</ul></li>
<li><strong>Introduction to <code>MetaboAnalyst</code> software</strong>
<ul>
<li>Overview of a useful, R-based resources for metabolomics</li>
<li>Illustrative workflow with <code>MetaboAnalyst</code></li>
</ul></li>
<li><strong>Elements of statistical Power Analysis</strong>
<ul>
<li>Brief review of hypothesis testing framework (from Module 2)</li>
<li>Review of <em>type I</em> and <em>type II</em> decision errors, contextualizing them in experimental settings</li>
<li>Understanding the test’s <em>statistical power</em> in connection to the <em>effect size</em> of an experiment</li>
</ul></li>
</ul>
</section>
<section id="programming-labs" class="level3" style="color:#77501a">
<h3 class="anchored" data-anchor-id="programming-labs"><span style="color:#005ca1"><i class="fa-brands fa-r-project" aria-label="r-project"></i></span> programming labs</h3>
<p>Each of the above module is accompanied by a matching practical session intended to consolidate the theoretical concepts via hands-on <strong>R coding</strong> sessions.</p>
<ul>
<li>The illustrative examples used are based on real biology/clinical research data.</li>
<li>In each example, the student is guided through the entire process: acquiring and reading data into R, identifying the appropriate analytical method, running the analysis and, finally, interpreting the obtained outcomes.</li>
<li>For every exercise, the participant is provided with input datasets (complete with documentation) and R code source files (<code>*.R</code>) with solved examples for future reference.</li>
<li>The instructor will also use the lab sessions as an opportunity to discuss common questions and challenges that are normally encountered in the day-to-day life of research scientists.</li>
</ul>
</section>
<!-- ### Module 1: Data cleaning and exploration review -->
<!-- + Introduction to end-to-end analysis (overview of the entire analysis process) -->
<!-- + Discussion of different data types (continuous, count, binary, multi-categorical) -->
<!-- + Discussion of the importance of “tidy data” when performing statistical analyses -->
<!-- + Introduction to advanced data cleaning methods -->
<!--  + Discussion of methods from the perspective of finding patterns in order to plan out your analyses and pre-processing steps -->
<!-- ### Module 1 Lab: -->
<!-- + Data wrangling 101: cleaning and structuring messy datasets -->
<!-- + Exploratory analysis and pre-processing, including visualization, outlier detection, and dimensionality reduction -->
<!-- + Identifying variables to be used in future analyses -->
<!-- ### Module 2: Dealing with Missingness -->
<!-- + Introduction to different kinds of missingness -->
<!-- + Discussion of how missingness impacts statistical analyses -->
<!-- + Review of complete-case analysis and its limitations -->
<!-- + Introduction to common types of imputation methods, including multiple imputation methods -->
<!-- + Discussion of when to use each type of method and the advantages and disadvantages of each -->
<!-- ### Module 2 Lab: -->
<!-- + Explore and visualize missingness to understand how it influences the analysis if not dealt with properly -->
<!-- + Impute missing data with single and multiple imputation methods (i.e. mean replacement, k-nearest neighbour, missForest, MICE) and use the imputed dataset(s) in analyses -->
<!-- ### Module 3: Modeling part 1 -->
<!-- + Overview of regression based statistical modeling methods: -->
<!--   + Regression (multiple linear regression models, generalized linear models for binary outcomes, count outcomes, binomial outcomes, etc.) -->
<!--   + Variable selection (lasso, elastic net) -->
<!-- + Model evaluation and selection -->
<!-- + Model visualization -->
<!-- ### Module 3 Lab: -->
<!-- + Identify the proper model given a particular question of interest with examples including -->
<!--   + Multiple linear regression analyses -->
<!--   + Generalized linear models for a diverse range of outcomes (e.g., binary, count, binomial) -->
<!-- + Conduct variable selection using regularization methods such as lasso and elastic net. -->
<!-- ### Module 4: Modeling part 2 -->
<!-- + Overview of non-regression based statistical modeling methods -->
<!--   + Classification (e.g., Random Forest) -->
<!--   + Clustering (K nearest neighbours) -->
<!-- + Model evaluation and selection -->
<!-- + Model visualization -->
<!-- ### Module 4 Lab: -->
<!-- + Identify appropriate classification or clustering techniques given a particular question of interest with examples including -->
<!--   + Logistic regression -->
<!--   + Random forest -->
<!--   + K-nearest neighbours -->
<!--   + Hierarchical clustering -->
<!-- + Evaluate and visualize chosen models.  -->
<!-- ### Module 5: Putting it all together -->
<!-- + Brief review of first four modules -->
<!-- + Finalizing a decision tree to be used as a tool when planning out your analysis -->
<!-- + Examples of identifying which model to use based on different datasets using the decision tree -->
<!-- ### Module 5 Lab: -->
<!-- + Discuss when the decision tree may not be sufficient for identifying a model and resources for continual learning. -->
<!-- + End-to-end analysis of real-world data starting with data cleaning, exploration, analysis, and interpretation -->
<!-- + Select from a list of curated datasets and address predefined research questions using the learned methods -->
<!-- ### Module 6: Introduction to Causal inference -->
<!-- + Causation vs. correlation -->
<!-- + Understanding the potential outcome framework -->
<!-- + How to estimate propensity scores with methods learned and use them for estimating the average treatment effect -->
<!-- ### Module 6 Lab: -->
<!-- + Assess the balance of covariates between treated and untreated groups -->
<!-- + Propensity score estimation via classification methods such as logistic regression and random forest -->
<!-- + Estimation of Average Treatment Effect (ATE) using propensity score matching and inverse weighting -->


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p><i class="fa-brands fa-creative-commons" aria-label="creative-commons"></i> <i class="fa-brands fa-creative-commons-by" aria-label="creative-commons-by"></i> <i class="fa-brands fa-creative-commons-nc" aria-label="creative-commons-nc"></i> <i class="fa-brands fa-creative-commons-sa" aria-label="creative-commons-sa"></i> <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/">CC BY-NC-SA 4.0</a> <br> M. Chiara Mimmi &amp; Luisa M. Mimmi 2024</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
<p>Made with <i class="fa-brands fa-r-project" aria-label="r-project"></i> and <a href="https://quarto.org/">Quarto</a></p>
</div>
  </div>
</footer>




</body></html>